[RFC] vim-patch:7.4.1975,7.4.1976,7.4.1977,7.4.1978,7.4.1979,7.4.1986,7.4.2029,7.4.2224,8.0.0219

jamessan opened the pull request:
#### vim-patch:7.4.1975

Problem:    On MS-Windows large files (> 2Gbyte) cause problems.
Solution:   Use "off_T" instead of "off_t".  Use "stat_T" instead of "struct
            stat".  Use 64 bit system functions if available.  (Ken Takata)

https://github.com/vim/vim/commit/8767f52fbfd4f053ce00a978227c95f1d7d323fe

Only the off_T changes are relevant, since all the "struct stat" usage
is abstracted by libuv.


#### vim-patch:7.4.1976

Problem:    Number variables are not 64 bits while they could be.
Solution:   Add the num64 feature. (Ken Takata)

https://github.com/vim/vim/commit/22fcfad29276bd5f317faf516637dcd491b96a12


#### vim-patch:7.4.1977

Problem:    With 64 bit changes don't need three calls to sprintf().
Solution:   Simplify the code, use vim_snprintf(). (Ken Takata)

https://github.com/vim/vim/commit/bde9810d6103ffe3a22a9330021cb21db1ed1792

nvim already had the equivalent code, so only the patch number was
needed.


#### vim-patch:7.4.1978

Problem:    Large file test does not delete its output.
Solution:   Delete the output. Check size properly when possible. (Ken Takata)

https://github.com/vim/vim/commit/c5af40ae646ceda817eff93b4f9ba274f031bea6


#### vim-patch:7.4.1979

Problem:    Getting value of binary option is wrong. (Kent Sibilev)
Solution:   Fix type cast.  Add a test.

https://github.com/vim/vim/commit/2acfbed9dbea990f129535de7ff3df360365130b


#### vim-patch:7.4.1986

Problem:    Compiler warns for loss of data.
Solution:   Use size_t instead of int. (Christian Brabandt)

https://github.com/vim/vim/commit/fef524bbff9aa186838c35212b2f89f61d627cf8

Equivalent change had already been made when merging earlier pack
patches.


#### vim-patch:7.4.2029

Problem:    printf() does not work with 64 bit numbers.
Solution:   use the "L" length modifier. (Ken Takata)

https://github.com/vim/vim/commit/38ee6b041e73ad31c3b6b99d56d20833b59b2b57


#### vim-patch:7.4.2224

Problem:    Compiler warnings with older compiler and 64 bit numbers.
Solution:   Add "LL" to large values. (Mike Williams)

https://github.com/vim/vim/commit/af9c4c9b5761c4c074237d87e2c95713bf721eab

Equivalent change was made in ZyX's typval refactoring.


#### vim-patch:8.0.0219

Problem:    Ubsan reports errors for integer overflow.
Solution:   Define macros for minimum and maximum values.  Select an
            expression based on the value. (Mike Williams)

https://github.com/vim/vim/commit/7a40ea2138102545848ea86a361f1b8dec7552b5
- [ ] ~~UBSAN failures~~ Ignored the remaining cases, as upstream is implicitly [doing].

[doing]: https://groups.google.com/forum/#!topic/vim_dev/_tqf8eQy5eA

jamessan committed:
vim-patch:7.4.1975

Problem:    On MS-Windows large files (> 2Gbyte) cause problems.
Solution:   Use "off_T" instead of "off_t".  Use "stat_T" instead of "struct
            stat".  Use 64 bit system functions if available.  (Ken Takata)

https://github.com/vim/vim/commit/8767f52fbfd4f053ce00a978227c95f1d7d323fe

Only the off_T changes are relevant, since all the "struct stat" usage
is abstracted by libuv.

jamessan committed:
vim-patch:7.4.1976

Problem:    Number variables are not 64 bits while they could be.
Solution:   Add the num64 feature. (Ken Takata)

https://github.com/vim/vim/commit/22fcfad29276bd5f317faf516637dcd491b96a12

jamessan committed:
vim-patch:7.4.1977

Problem:    With 64 bit changes don't need three calls to sprintf().
Solution:   Simplify the code, use vim_snprintf(). (Ken Takata)

https://github.com/vim/vim/commit/bde9810d6103ffe3a22a9330021cb21db1ed1792

nvim already had the equivalent code, so only the patch number was
needed.

jamessan committed:
vim-patch:7.4.1978

Problem:    Large file test does not delete its output.
Solution:   Delete the output. Check size properly when possible. (Ken Takata)

https://github.com/vim/vim/commit/c5af40ae646ceda817eff93b4f9ba274f031bea6

jamessan committed:
vim-patch:7.4.1979

Problem:    Getting value of binary option is wrong. (Kent Sibilev)
Solution:   Fix type cast.  Add a test.

https://github.com/vim/vim/commit/2acfbed9dbea990f129535de7ff3df360365130b

jamessan committed:
vim-patch:7.4.1986

Problem:    Compiler warns for loss of data.
Solution:   Use size_t instead of int. (Christian Brabandt)

https://github.com/vim/vim/commit/fef524bbff9aa186838c35212b2f89f61d627cf8

Equivalent change had already been made when merging earlier pack
patches.

jamessan committed:
vim-patch:7.4.2029

Problem:    printf() does not work with 64 bit numbers.
Solution:   use the "L" length modifier. (Ken Takata)

https://github.com/vim/vim/commit/38ee6b041e73ad31c3b6b99d56d20833b59b2b57

justinmk commented:
LGTM. QB failure:

```
14:47:51,043 WARN  - /Users/quickbuild/tmp/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/eval/decode.c:365:55: error: incompatible pointer types passing 'unsigned long *' to parameter of type 'uvarnumber_T *' (aka 'unsigned long long *') [-Werror,-Wincompatible-pointer-types]
14:47:51,043 WARN  - STR2NR_HEX | STR2NR_FORCE, NULL, &ch, 4);
14:47:51,043 WARN  - ^~~
14:47:51,043 WARN  - /Users/quickbuild/tmp/buildagent/workspace/root/neovim/pull-requests-automated/build/Release/include/charset.h.generated.h:59:138: note: passing argument to parameter 'unptr' here
14:47:51,043 WARN  - void vim_str2nr(const char_u *const start, int *const prep, int *const len, const int what, varnumber_T *const nptr, uvarnumber_T *const unptr, const int maxlen);
14:47:51,043 WARN  - ^
14:47:51,049 WARN  - 1 error generated.
```

jamessan commented:
`msgpack_to_vim` and `TYPVAL_ENCODE_DEFINE_CONV_FUNCTIONS` need an update to handle `kMPInteger` conversions since `varnumber_T` is now 64-bit.

Shougo commented:
Vim 7.4.1980 is also needed.
It is the additional fix of 7.4.1979.

jamessan commented:
> Vim 7.4.1980 is also needed.

Why? That's a performance improvement for quickfix.

Shougo commented:
I had mistake.... Sorry.

bfredl commented:
roundtriping 64-bit numbers seems to work well both with rpc (tested with `:python3`/`py3eval`) and `msgpackdump`/`parse`. Maybe add tests for these?

justinmk commented on src/nvim/eval/decode.c:
```diff
@@ -965,9 +964,8 @@ int msgpack_to_vim(const msgpack_object mobj, typval_T *const rettv)
-        list_append_number(list, (varnumber_T) ((n >> 31) & 0x7FFFFFFF));
-        list_append_number(list, (varnumber_T) (n & 0x7FFFFFFF));
+        list_append_number(list, (varnumber_T)((n >> 63) & 0x1));
+        list_append_number(list, (varnumber_T)(n & 0x7fffffffffffffff));
```
will these changes break shada back-compat?

bfredl replied:
old shada files should only contain 32-bit msgpack values (where a typval is encoded/decoded), which will be interpreted the same. New shada files with 64-bit values will be missinterpreted by old nvim (values get truncated), of course. (maybe better an error, but, hard to change the code of nvim executables out in the wild...) But 32-bit values in new shada files will still be interpreted correctly by old nvim, so it is "as compatible it possibly could be".

justinmk commented:
Which commit did the UBSAN failures start with?

jamessan commented:
> Which commit did the UBSAN failures start with?

7.4.1976.  Vim has the same UBSAN failures (plus some that I've fixed in this PR, but am unsure if it's the right way).

The core of it comes down to 64-bit math over/underflowing the bounds of a signed 64-bit variable, which is undefined.  In a few places, I've used unsigned variables so it's not UB, but I'm not sure what the expected behavior that nvim should be exhibiting is.

* Do we add checks to clamp to the signed range (which is fairly straight forward for `+`/`-`, but less so for other operations)?
* Do we force wrapping for signed arithmetic (`-fwrapv`)?
* Do we use language extensions to perform the operations using 128-bit variables and then assign the value to the final 64-bit variable?
* Do we just ignore it, like Vim does, and say "You were trying to do math outside the range of the variables" if people complain (which means we would need to tell UBSAN to ignore the relevant errors)?

If it were the last one, then 89c2254 could be removed.

jamessan commented:
The UB with regard to 64-bit math is now being discussed [upstream] too.  Hopefully, something comes out of that soon so we can wrap this up.

[upstream]: https://www.mail-archive.com/vim_dev@googlegroups.com/msg43006.html

jamessan committed:
vim-patch:8.0.0219

Problem:    Ubsan reports errors for integer overflow.
Solution:   Define macros for minimum and maximum values.  Select an
            expression based on the value. (Mike Williams)

https://github.com/vim/vim/commit/7a40ea2138102545848ea86a361f1b8dec7552b5

jamessan committed:
lint

jamessan committed:
vim-patch:7.4.2224

Problem:    Compiler warnings with older compiler and 64 bit numbers.
Solution:   Add "LL" to large values. (Mike Williams)

https://github.com/vim/vim/commit/af9c4c9b5761c4c074237d87e2c95713bf721eab

Equivalent change was made in ZyX's typval refactoring.

jamessan commented:
Raised the issue of defining behavior for overflow/underflow on vim-dev -- https://groups.google.com/forum/#!topic/vim_dev/_tqf8eQy5eA.

jamessan commented:
Given Bram's response, I think I'm just going to add suppressions for now.  If anyone actually steps up to handle the UB, we'll merge it.

jamessan committed:
func_attr: Allow disabling UBSAN for a function

jamessan committed:
*: Disable UBSAN for VimL arithmetic implementation

After merging +num64, the 64-bit sanitizer builds show that Vim doesn't
buffer the user from C's UB in signed arithmetic.  Upstream doesn't
appear to be [interested] in fixing the issue, so suppress UBSAN until
someone decides to fix the problem.

N.B., the problem existed before but went unnoticed since the sanitizer
builds weren't being run in 32-bit mode.

[interested]: https://groups.google.com/d/topic/vim_dev/_tqf8eQy5eA/discussion

jamessan committed:
eexe_mod_op: Explicitly cast varnumber_T to float_T for -Wconversion

oni-link commented on src/nvim/eval.c:
```diff
@@ -4023,18 +4023,20 @@ eval6 (
+        } else if (op == '/') {
+          if (n2 == 0) {                // give an error message?
+            if (n1 == 0) {
+              n1 = VARNUMBER_MIN;  // similar to NaN
```
```
:echo 0/0
-9223372036854775808 <- Help says this is NaN
:eval (0/0)
-9223372036854775807 <- Help says this is negative infinity
```

jamessan replied:
Note that Vim's help doesn't reflect reality, or I'm misunderstanding it.  Using 8.0.0600:

```
When 64-bit Number support is enabled:
	  0 / 0  = -0x8000000000000000	(like NaN for Float)
	 >0 / 0  =  0x7fffffffffffffff	(like positive infinity)
	 <0 / 0  = -0x7fffffffffffffff	(like negative infinity)
```

However:

```
:echo printf('%d %#x', 0/0, 0/0)
-9223372036854775808 0x8000000000000000
:echo printf('%d %#x', 1/0, 1/0)
9223372036854775807 0x7fffffffffffffff
:echo printf('%d %#x', -1/0, -1/0)
-9223372036854775807 0x8000000000000001
:echo printf('%d %d', 0/0, eval(0/0))
-9223372036854775808 -9223372036854775807
```

so we're at least consistent with Vim's inconsistency.

jamessan commented:
@ZyX-I or @bfredl, could one of you finish this PR?  I think I've taken this as far as I can.  The tests for round-tripping edge-case values are failing and @oni-link's [comment](https://github.com/neovim/neovim/pull/5621#discussion_r115136849) probably needs looking into.

jamessan committed:
Merge remote-tracking branch 'origin/master' into vim-7.4.1975

jamessan committed:
vim-patch:8.0.0614

Problem:    float2nr() is not exactly right.
Solution:   Make float2nr() more accurate.  Turn test64 into a new style test.
            (Hirohito Higashi, closes vim/vim#1688)

https://github.com/vim/vim/commit/863e80b4451b5102b41bebf9ddca3a420de746fa

jamessan committed:
eval/typval: Convert string to varnumber_T instead of intermediate long

jamessan committed:
strings/tv_float: Explicitly cast v_number to float_T for -Wconversion

jamessan commented:
@neovim/core The last problem to getting this merged is this failure:

```
[  FAILED  ] ...m/neovim/test/functional/eval/msgpack_functions_spec.lua @ 347: msgpack*() functions can restore and dump INT64_MIN
...m/neovim/test/functional/eval/msgpack_functions_spec.lua:354: Expected objects to be the same.
Passed in:
(number) 0
Expected:
(number) 1
stack traceback:
	...m/neovim/test/functional/eval/msgpack_functions_spec.lua:354: in function <...m/neovim/test/functional/eval/msgpack_functions_spec.lua:347>
```

I'd appreciate any help fixing that.

justinmk commented:
Maybe it gives a clue to change the test from:

    eq(1, eval('-0x8000000000000000 == parsed[0]'))

to:

    eq(-8000000000000000, eval('parsed[0]'))

jamessan commented:
That's marshaling the data back through msgpack + lua, which isn't what we're testing.  I can change it to an `assert_equal()` call instead and then check `v:errors` in lua.

justinmk commented:
Ah, ok. I also noticed that before this PR, the `else` branch of that test was taken. Is that expected?

`eval('string(parsed[0])')` returns `-9223372036854775808` ...

jamessan commented:
> I also noticed that before this PR, the else branch of that test was taken. Is that expected?

I think so, but @ZyX-I would likely know for sure.  If I'm understanding it right, the else branch is used when Vim can't represent the number natively so it fallsback to the structured type.

jamessan committed:
functests/msgpack: Use assert_equal() for more informative errors

ZyX-I commented on test/functional/eval/msgpack_functions_spec.lua:
```diff
@@ -351,7 +352,8 @@ describe('msgpack*() functions', function()
                '|| parsed[0]._TYPE is v:msgpack_types.integer'))
     if eval('type(parsed[0]) == type(0)') == 1 then
-      eq(1, eval('-0x8000000000000000 == parsed[0]'))
+      command('call assert_equal(-0x8000000000000000, parsed[0])')
```
This looks like a bug somewhere. Why `assert_equal()` works, but `==` does not? But @jamessan is right, `else` branch is for the case when number does not fit into `[VARNUMBER_MIN, VARNUMBER_MAX]`.

jamessan replied:
`==` works but it doesn't provide an informative error when it fails.  That's why I switched it to `assert_equal()`, so we get information back from `v:errors`.

ZyX-I replied:
Better use `string()` and `eq()`, it is highly uncommon in non-legacy lua tests to use parts of Vim own testing infrastructure.

ZyX-I commented on test/functional/eval/msgpack_functions_spec.lua:
```diff
@@ -337,7 +337,8 @@ describe('msgpack*() functions', function()
                '|| parsed[0]._TYPE is v:msgpack_types.integer'))
     if eval('type(parsed[0]) == type(0)') == 1 then
-      eq(1, eval('0xFFFFFFFFFFFFFFFF == parsed[0]'))
+      command('call assert_equal(0xFFFFFFFFFFFFFFFF, parsed[0])')
```
And this looks overly strange: UINT64_MAX is *unsigned* integer maximum, it *must not* fit into one `varnumber_T` if it is `int64_t` and not a 128-bit integer available on some systems. If test fails because it entered this branch then there is a bug in `msgpackparse()`.

ZyX-I replied:
Wondering whether I needed casts in

```C
      if (mobj.via.u64 <= VARNUMBER_MAX) {
…
      if (mobj.via.i64 >= VARNUMBER_MIN) {
 ```

? AFAIR that should be upcasted automatically because `int64_t` rank is greater then `int` (equal with 64-bit `varnumber_T`), `uint64_t` is greater then `int` and `int64_t` should be promoted to `uint64_t` when comparing.

jamessan replied:
It's not entering this branch.  The uint64 test is properly falling into the else.

ZyX-I commented on src/nvim/eval/typval.h:
```diff
@@ -20,20 +20,21 @@
 /// Maximal possible value of varnumber_T variable
-#define VARNUMBER_MAX INT_MAX
+#define VARNUMBER_MAX INT64_MAX
+#define UVARNUMBER_MAX UINT64_MAX
```
Missing documentation.

jamessan commented:
With the change to `assert_equal()`, the failure now explicitly shows us the same problem that @justinmk derived.

```
[ RUN      ] msgpack*() functions can restore and dump INT64_MIN: FAIL
...m/neovim/test/functional/eval/msgpack_functions_spec.lua:356: Expected objects to be the same.
Passed in:
(table) {
 *[1] = 'Expected -9223372036854775807 but got -9223372036854775808' }
Expected:
(table) { }
stack traceback:
	...m/neovim/test/functional/eval/msgpack_functions_spec.lua:356: in function <...m/neovim/test/functional/eval/msgpack_functions_spec.lua:348>
```

ZyX-I commented:
> Ah, ok. I also noticed that before this PR, the else branch of that test was taken. Is that expected?

> eval('string(parsed[0])') returns "-9223372036854775808" ...

> Edit: string(-0x8000000000000000) => -9223372036854775807. So the string(parsed[0]) result is off by one.

First number (`…8`) is even, `string(-0x8…0)` is odd. Which one is wrong is obvious. Did you try first test from lua (e.g. `funcs.string(-0x8…0)`)? Doing this is entirely incorrect, lua does not use integers normally, `lua_Number` is `double`.

ZyX-I commented:
@jamessan So it is not “lua double” problem, but a problem with VimL numbers parser?

jamessan commented:
> So it is not “lua double” problem, but a problem with VimL numbers parser?

Yeah, it's definitely a VimL problem.

ZyX-I commented:
@jamessan Still I would write `eq('-0x8…0', eval('printf("0x%x", parsed[0])'))`. Or similar, but with `string()` and not `printf()`.

ZyX-I commented:
And a separate test for `eval('-0x8…0')` somewhere else.

jamessan commented:
Feel free to push some commits to my branch. :)

ZyX-I commented:
Though, forgot a thing: `%x` is for unsigned integrals. This only leaves `string()`.

jamessan commented:
Ok, it looks like the problem is that `vim_str2nr` (called from `eval7`) isn't getting the leading `-` for the number, so it then converts `0x8000000000000000` to `VARNUMBER_MAX` (`0x7fffffffffffffff`).  Once we get out of `vim_str2nr`, `eval7` then handles the unary minus which negates the value, resulting in `-9223372036854775807` (or `0x8000000000000001`).

jamessan commented:
Vim itself _used_ to handle this correctly (circa 8.0.0197) but doesn't anymore...

jamessan committed:
functests/msgpack: Correct representation of literal INT64_MIN

In order to generate INT64_MIN from literal values, it's necessary to
use "-0x7fffffffffffffff - 1".  Using "-0x8000000000000000" causes the
value to get clamped to INT64_MAX and then negated.

jamessan commented:
Ok, this is the same issue as with C.  You can't represent `0x80000000000000` as a positive value, because that's beyond INT64_MAX, so it gets clamped to `0x7fffffffffffffff` and then negated.  There was a similar fix in Vim's tests when [8.0.0219] was applied.

[8.0.0219]: https://github.com/vim/vim/releases/tag/v8.0.0219

jamessan committed:
*: Fix conversion warnings for tv_get_number*()

jamessan committed:
lint
